nvflare/app_opt/quantization/model_dequantizor.py

```diff
@@ -0,0 +1,117 @@
+# Copyright (c) 2024, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Union
+
+import numpy as np
+import torch
+from bitsandbytes.functional import dequantize_blockwise
+
+from nvflare.apis.dxo import DXO, DataKind, MetaKey
+from nvflare.apis.dxo_filter import DXOFilter
+from nvflare.apis.fl_context import FLContext
+from nvflare.apis.shareable import Shareable
+from nvflare.app_opt.quantization.constant import DATA_TYPE, QUANTIZATION_TYPE
+
+
+class ModelDequantizor(DXOFilter):
+    def __init__(self, source_data_type="float32"):
+        """Filter to dequantize Shareable object to recover from quantization
+
+        Args:
+            source_data_type: original data type of the model
+
+        """
+
+        # support weight and weight_diff data kinds
+        data_kinds = [DataKind.WEIGHTS, DataKind.WEIGHT_DIFF]
+        super().__init__(supported_data_kinds=data_kinds, data_kinds_to_filter=data_kinds)
+
+        # assign data type and check if it is valid
+        self.logger.info("Using model dequantizator.")
+        if source_data_type.upper() not in DATA_TYPE:
+            raise ValueError(f"Invalid source data type: {source_data_type}")
+        else:
+            self.source_data_type = source_data_type
+
+    def dequantization(self, params: dict, quant_state: dict, fl_ctx: FLContext):
+        n_params = len(params.keys())
+        self.log_info(fl_ctx, f"Running dequantization on {n_params} variables")
+        n_bytes_before = 0
+        n_bytes_after = 0
+        n_bytes_meta = 0
+        for i, param_name in enumerate(params.keys()):
+            if self.source_data_type == "float32":
```

Comment:
Seems this is the only supported condition. Why check for it?

Reply:
for future-proof, potentially we can add more

Reply:
let's check the precision for each sub-vec in case of mixed precision

Reply:
and tensor / other state_dict / numpy

Reply:
check the param type and only perform quantization to numpy float32, will handle tensor for later when we enable transfer tensor directly